models/src/main/scala/coop/rchain/models/ParSet.scala

```diff
@@ -36,13 +36,23 @@ object ParSet {
 
   def apply(
       ps: Seq[Par],
-      connectiveUsed: Boolean = false,
-      remainder: Option[Var] = None
+      remainder: Option[Var]
   ): ParSet = {
     val shs = SortedParHashSet(ps)
-    ParSet(shs, connectiveUsed, Coeval.delay(updateLocallyFree(shs)).memoize, remainder)
+    ParSet(
+      shs,
+      connectiveUsed(ps) || remainder.isDefined,
+      Coeval.delay(updateLocallyFree(shs)).memoize,
+      remainder
+    )
   }
 
-  def updateLocallyFree(ps: SortedParHashSet): BitSet =
+  def apply(ps: Seq[Par]): ParSet =
+    apply(ps, None)
+
+  private def connectiveUsed(seq: Seq[Par]): Boolean =
+    seq.exists(_.connectiveUsed)
```

Comment:
We actually need the field, or we'll rely on our children's caches - which is a terrible thing for a parent to do

Reply:
Also, we should probably make the `connectiveUsed` computation lazy... At this point I think we should maybe allow passing it from outside as it used to be?

Reply:
> We actually need the field, or we'll rely on our children's caches - which is a terrible thing for a parent to do

what caches? I think this version is safer than it was before.

Reply:
It is and I'd very much welcome the change. `connectiveUsed` exists on the tree so that it doesn't have to be re-evaluated each time we ask about it. Ohterwise we'd just use the HasLocallyFree typeclass.